src/runtime_src/core/tools/xbutil2/OO_MemWrite.cpp

```diff
@@ -104,49 +115,145 @@ XBU::verbose("SubCommand option: read mem");
+      // tellg returns a signed value as the number of bytes. Validate the return code and
+      // cast it into a useful format
+      auto nonvalidated_length = input_stream.tellg();
+      if (nonvalidated_length < 0)
+        throw std::runtime_error("Failed to get input file length");
+      uint64_t validated_length = static_cast<uint64_t>(nonvalidated_length);
+
+      if (m_sizeBytes.empty()) // update size
+        size = validated_length;
+      count = static_cast<int>(std::ceil(validated_length / size));
+      input_stream.seekg(0, input_stream.beg);
+    }
+    else
+      count = m_count;
+
+    if (count <= 0)
+      throw xrt_core::error(std::errc::operation_canceled, "Value for --count must be greater than 0");
+
+    // Logging information
+    XBU::verbose(boost::str(boost::format("Device: %s") % xrt_core::query::pcie_bdf::to_string(xrt_core::device_query<xrt_core::query::pcie_bdf>(device))));
+    XBU::verbose(boost::str(boost::format("Address: %s") % addr));
+    XBU::verbose(boost::str(boost::format("Size: %llu") % size));
+    XBU::verbose(boost::str(boost::format("Block count: %d") % count));
+    XBU::verbose(boost::str(boost::format("Input File: %s") % m_inputFile));
+    XBU::verbose(boost::str(boost::format("Bytes to write: %lld") % 5));
+
+    // Write to device memory
+    XBU::xclbin_lock xclbin_lock(device);
+    
+    try {
+      for (int c = 0; c < count; c++) {
+        XBU::verbose(boost::str(boost::format("[%d / %d] Writing to Address: %s, Size: %llu bytes") % c % count % addr % size));
+        std::vector<char> buffer(size);
+        // gcount will only return a value >= 0
+        uint64_t input_size = static_cast<uint64_t>(input_stream.read(buffer.data(), size).gcount());
+        xrt_core::device_mem_write(device.get(), addr, buffer);
+        if (input_size != size)
+          break; // partial read and break the loop
+        addr += size;
+      }
+    } catch (const xrt_core::error& e) {
+      std::cerr << e.what() << std::endl;
+      return;
+    }
+    std::cout << "Memory write succeeded" << std::endl;
+
+    return;
+  }
+
+  // Parse the fill option path
   if(!m_fill.empty()) {
+    // Validate the block count. This cannot be done as the --input path may create its own block count
+    if (m_count <= 0)
+      throw xrt_core::error(std::errc::operation_canceled, "Value for --count must be greater than 0");
+    
+    // Validate the number of bytes to write
+    if (m_sizeBytes.empty())
+      throw xrt_core::error(std::errc::operation_canceled, "Value required for --size when using --fill");
+
+    char fill_byte = 'J';
     try {
       //-- fill pattern
-      fill_byte = std::stoi(m_fill, nullptr, 0);
+      fill_byte = static_cast<char>(std::stoi(m_fill, nullptr, 0));
     }
     catch(const std::invalid_argument&) {
       std::cerr << boost::format("ERROR: '%s' is an invalid argument for '--fill'. Please specify a value between 0 and 255\n") % m_fill;
```

Comment:
What happens if the architecture has a `signed char`?

Reply:
Regardless of the representation, the byte is fills a vector for writing. At the end of the day even if it is an unsigned char all we really care about is the value of the variable not the type.

Reply:
If you have an `unsigned char` I do not have a problem. If you have a `signed char` this is undefined behavior to have an overflow.
I guess this is not the major problem of XRT anyway...

Reply:
Doesn't the value just become negative? For example: value = 128, unsigned char => 128, char => -1. At the end of the day the data that is written into the memory is still 0x10. Most of the read functions available require a char* to populate so changing this would be difficult at this level. For the Memory read function I definitely see your point and that should be changed to a uint8_t.